### DISCUSSION TITLE
Run on fan less mini PC with GPU

### DISCUSSION THREAD

**Discussion by koh-osug** [openai/whisper#417] [Q&A]
The impressing results even better than any STT cloud services expect one make this a good fit for an assistant application. I would like to tun this on a fan less mini PC with a GPU to avoid using some cloud STT services for privacy, billing and offline reasons. My ordinary PC without a GPU is too slow to translate an audio in real time, in fact it takes minutes. So I was wondering if a GPU powered device, maybe one of the Nvidia Jetson devices would work. Are there any recommendations out there? Since the device is running 24/7 a very low energy consumption and no mechanical parts, hence the fan less  design, would be great.

**Comment by Topping1** [comment]
try this C++ port, it is optimized for CPU: https://github.com/ggerganov/whisper.cpp

**Comment by koh-osug** [comment]
I have tried this solution, but for a 15s transcription it took 205 seconds with the large model. Well, the first part of the transcript was already finished after maybe 30s, which is still a lot, but then the end which is silence took the remaining time. The original Python solution is faster. Any ideas how to further improve this?

> whisper_print_timings:     load time =  2355.22 ms
> whisper_print_timings:      mel time =   115.76 ms
> whisper_print_timings:   sample time =     0.00 ms
> whisper_print_timings:   encode time = 194114.92 ms / 6066.09 ms per layer
> whisper_print_timings:   decode time =  8427.01 ms / 263.34 ms per layer
> whisper_print_timings:    total time = 205026.06 ms

**Reply by ggerganov** [reply]
The `whisper.cpp` implementation is optimized for Apple Silicon, so you should get much better performance using an Apple device. For example, on Mac M1 Mini the Encoder part of the model, using the `large` model currently takes about 7-8 seconds.

Regarding the blank audio processing - using some sort of VAD should help, but it is not yet supported. Maybe you can try to perform some external pre-processing first to isolate the speech and then pass the voice-only audio to Whisper.

**Reply by koh-osug** [reply]
I assume I can add also some other compiler flags to optimize the binary for a different architecture? Or are you using some specific assembler instructions in the code to prevent this?  A Mac M1 Mini is too expensive and AFAIK it cannot run Linux.

**Comment by richardburleigh** [comment]
You may want to check out something like the[ Intel Neural Compute Stick](https://www.intel.com/content/www/us/en/developer/tools/neural-compute-stick/overview.html) to offload the inference.

I don't own one so you would need to do more research about it.

Edit: There are some other options [here](https://pallavaggarwal.in/ai-hardware-platforms/).

**Reply by kaoh** [reply]
The Intel compute stick is only using an Atom processor. How can this accelarate the inference if not even my Intel i7 can do it near realtime? I will check the other options.

**Reply by kaoh** [reply]
Oh, I mixed it up. There is a Intel neural compute stick. But this product is discontinued. Not even sure if PyTorch is supporting it.

**Comment by koh-osug** [comment]
I have bought a Nvidia Xavier 16 GB based computer. The setup is not straightforward since a special pytorch version is needed. Once this is installed I could use a medium model and the transcription for a 4 s audio file took 8 s. Is this expected? I though the Xavier is pretty fast. I could also not load the large model. I always get as output on the command line "Killed", it seems it cannot be loaded. I though 16 GB RAM are enough in the GPU?

**Reply by kaoh** [reply]
The Nvidia Xavier 16 GB based computer is from Seeed. It turned out that the 16 GB are shared memory, hence not the full memory can be used for a model. Also, the first transcription is slow because the network has to  be initialized. After sorting it this issues the speed is almost 1:1, i.e. 10s speech = 10s transcription. I'm searching now for some newer Mini computers supporting the latest Nvidia Orin versions.

**Reply by ysiegel29** [reply]
@koh-osug interesting thread, I wanted to try something similar on my jetson xavier NX. Did you get any further? What performance did you get with the medium model? It loaded fine? how long inference took? Thanks!

**Reply by koh-osug** [reply]
> @koh-osug interesting thread, I wanted to try something similar on my jetson xavier NX. Did you get any further? What performance did you get with the medium model? It loaded fine? how long inference took? Thanks!

The medium model works. Performance is approx. 1:1, i.e. 10s audio = 10s processing. This first transcription is slower. The new Orin NX version will come next year with 5x performance. But I fear it will be ridiculous expensive.

**Reply by ysiegel29** [reply]
Ok many thanks
